db/db_impl.cc

```diff
@@ -631,7 +634,86 @@ void DBImpl::StartTimedTasks() {
             stats_dump_period_sec * 1000000));
       }
     }
+    stats_persist_period_sec = mutable_db_options_.stats_persist_period_sec;
+    if (stats_persist_period_sec > 0) {
+      if (!thread_persist_stats_) {
+        thread_persist_stats_.reset(new rocksdb::RepeatableThread(
+            [this]() { DBImpl::PersistStats(); }, "pst_st", env_,
+            stats_persist_period_sec * 1000000));
+      }
+    }
+  }
+}
+
+void DBImpl::PersistStats() {
+  TEST_SYNC_POINT("DBImpl::PersistStats:1");
+#ifndef ROCKSDB_LITE
+  const DBPropertyInfo* cf_property_info =
+      GetPropertyInfo(DB::Properties::kCFStats);
+  assert(cf_property_info != nullptr);
+  const DBPropertyInfo* db_property_info =
+      GetPropertyInfo(DB::Properties::kDBStats);
+  assert(db_property_info != nullptr);
+  std::map<std::string, std::string> stats_map;
+  if (shutdown_initiated_) {
+    return;
+  }
+  {
+    InstrumentedMutexLock l(&mutex_);
+    default_cf_internal_stats_->GetMapProperty(
+        *db_property_info, DB::Properties::kDBStats, &stats_map);
+    for (auto cfd : *versions_->GetColumnFamilySet()) {
+      if (cfd->initialized()) {
+        cfd->internal_stats()->GetMapProperty(
+            *cf_property_info, DB::Properties::kCFStatsNoFileHistogram,
+            &stats_map);
+      }
+    }
+    for (auto cfd : *versions_->GetColumnFamilySet()) {
+      if (cfd->initialized()) {
+        cfd->internal_stats()->GetMapProperty(
+            *cf_property_info, DB::Properties::kCFFileHistogram, &stats_map);
+      }
+    }
   }
+  TEST_SYNC_POINT("DBImpl::PersistStats:2");
+  WriteOptions wo;
+  ColumnFamilyOptions cf_options;
+  const uint64_t now_micros = env_->NowMicros();
+  stats_history_[now_micros] = stats_map;
+  // delete older stats snapshots to control memory consumption
+  int num_stats_to_delete =
+      static_cast<int>(stats_history_.size()) -
+      static_cast<int>(mutable_db_options_.max_stats_history_count);
+  if (num_stats_to_delete > 0) {
+    auto it = stats_history_.begin();
+    for (int i = 0; i < num_stats_to_delete; ++i) {
+      if (it != stats_history_.end()) {
+        it = stats_history_.erase(it);
+      }
+    }
+  }
+  // TODO: also persist stats to disk
+#endif  // !ROCKSDB_LITE
+}
+
+std::map<uint64_t, std::map<std::string, std::string> >
+DBImpl::GetStatsHistory(GetStatsOptions& stats_opts) {
+  return FindStatsBetween(stats_opts.start_time, stats_opts.end_time);
+}
+
+std::map<uint64_t, std::map<std::string, std::string> >
+DBImpl::FindStatsBetween(uint64_t start_time, uint64_t end_time) {
+  std::map<uint64_t, std::map<std::string, std::string> > stats_history;
+  // first dump whatever is in-memory that satisfies the time range
+  for (const auto& stats : stats_history_) {
```

Comment:
Is `stats_history_` thread safe? Is it safe to access it while another thread might be updating it?